### ISSUE TITLE
Questions about Loss

### ISSUE THREAD

**Issue by jypjypjypjyp** [chenghao-shi/MDGAT-matcher#7]
Hi, Thank you for your fantastic project. 
I'm a little confused about gap loss or triplet loss. 
```
before_clamp_loss = loss_anc_pos - loss_anc_neg + self.triplet_loss_gamma
```
We know that both loss_anc_pos and loss_anc_neg are in a -log manifold, but the triplet_loss_gamma is a constant. When the score is low, the difference between positive loss and negative loss will become large, which may causes the triplet_loss_gamma to not work well.

There is an example:
![origin_img_v2_403ac63e-3f23-48ff-9399-11186fe28f1g](https://user-images.githubusercontent.com/34328687/165048891-aa35c520-6d49-4359-b18d-ffb2973a5c5f.jpg)
the positive score is 0.1, the negative score is 0.01, and  triplet_loss_gamma is set to 1. their loss equals -log(0.1) - (-log(0.01))+1=2.3-4.6+1<0
The loss will be ignored, but the negative score is only 0.09 worse than positive score.

The above is my confusion.

**Comment by chenghao-shi** [comment]
Hi，


Actually, if you double check the code, both gap loss and triplet loss are not in a log manifold, as the scores have been pre-processed by exp().
Similar operation can also be observed in circle loss using logsumexp().


The problem you describe is one that we need to avoid when designing a loss.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "chenghao-shi/MDGAT-matcher"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年4月25日(星期一) 下午4:23
***@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;[chenghao-shi/MDGAT-matcher] Questions about Loss (Issue #7)





 
Hi, Thank you for your fantastic project.
 I'm a little confused about gap loss or triplet loss.
 before_clamp_loss = loss_anc_pos - loss_anc_neg + self.triplet_loss_gamma  
We know that both loss_anc_pos and loss_anc_neg are in a -log manifold, but the triplet_loss_gamma is a constant. When the score is low, the difference between positive loss and negative loss will become large, which may causes the triplet_loss_gamma to not work well.
 
There is an example:
 
 the positive score is 0.1, the negative score is 0.01, and  triplet_loss_gamma is set to 1. their loss equals -log(0.1) - (-log(0.01))+1=2.3-4.6+1<0
 The loss will be ignored, but the negative score is only 0.09 worse than positive score.
 
The above is my confusion.

**Comment by jypjypjypjyp** [comment]
Your answer solved my question. Thanks a lot!
